Warehouse tab widget

RichardMXt opened the pull request.

RichardMXt committed:
Basic resource amount display and storage limit

RichardMXt committed:
Tab widget display fixes when Multi-selecting
*Reseting progress bar when multiple building are selected
*Deleted "if same classes selected" check

MichaelXt commented on Assets/UI/BasicControls/InventorySlot2.gd:
```diff
@@ -0,0 +1,45 @@
+    return resource_type
+  set(value):
+    resource_type = value
+    if self.is_node_ready() == false:
```
change to check to resource_image != null. Do the setting in _ready() too.

MichaelXt commented on Assets/UI/TabWidgets/ViewComponents/ProductionChain.gd:
```diff
@@ -44,15 +44,18 @@ func _ready():
         var production_time = building.building_data.processing_time
         progress = (production_time - building.production_timer.time_left) / production_time
       else:
         progress = 0
```
set progress=0 before loop?

MichaelXt commented on Assets/UI/TabWidgets/WarehouseTabWidget.gd:
```diff
@@ -0,0 +1,16 @@
+    if len(self.selected_objects) == 1:
+      var resources: Dictionary = GameStats.game_stats_resource.resources
+      # pass through all inventory slots and update them
+      for item_slot: InventorySlot2 in settlement_inventory.get_children():
```
move to function. the name would tell what this code does

MichaelXt commented on Assets/World/Buildings/Warehouse/Warehouse2D.gd:
```diff
@@ -33,10 +34,11 @@ func unload_worker(objects_to_unload: Dictionary) -> void:
-      GameStats.game_stats_resource.resources[object] += amount
+    var resource_amount = GameStats.game_stats_resource.resources.get(object)
+    if resource_amount != null:
+      GameStats.game_stats_resource.resources[object] = min(storage_capacity, resource_amount + amount)
```
add comment: // TODO: some resources will be thrown out. Should we notify user or carry them back?

RichardMXt committed:
Resource overlay and some bug and naming fixes

RichardMXt committed:
Resource pinning, pin selection, and get_node befor ready fixes

RichardMXt committed:
Increased display slot size and set a limit on select slot

RichardMXt committed:
Some bug fixes and code clean-up
*made "can update info check" check if has the nodes not self.is_ready
*avoided "else" code, by setting defult vals of vars
*moved code into func to clean-up code and ease understanding
*added todo note for notifing user
